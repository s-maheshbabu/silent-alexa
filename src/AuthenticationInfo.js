import { hasIn } from "immutable";
import util from "util";

export default class AuthenticationInfo {
  constructor(authResponse) {
    if (arguments.length == 1 && this.validateAuthResponse(authResponse)) {
      var _access_token = authResponse.access_token;
      var _expires_in = authResponse.expires_in;
    }
    this.getAccessToken = function() {
      return _access_token;
    };
  }

  isValid() {
    return this.getAccessToken() ? true : false;
  }

  validateAuthResponse(authResponse) {
    if (
      hasIn(authResponse, ["access_token"]) &&
      hasIn(authResponse, ["expires_in"]) &&
      authResponse.access_token &&
      authResponse.expires_in
    ) {
      return true;
    } else {
      console.log(
        "Encountered an error on login: " +
          util.inspect(authResponse, { showHidden: true, depth: null })
      );
      return false;
    }
  }
}
