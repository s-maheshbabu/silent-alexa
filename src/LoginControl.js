import React from "react";
import LoginButton from "./LoginButton";

// Options variable to request for implicit grant.
// TODO: Logic for assigning 'deviceSerialNumber' needs to be revisited.
const options = Object.freeze({
  scope: ["alexa:all", "profile"],
  scope_data: {
    "alexa:all": {
      productID: "Silent_Alexa",
      productInstanceAttributes: { deviceSerialNumber: "12345" }
    }
  }
});
export default class LoginControl extends React.Component {
  render() {
    return <LoginButton onClick={() => this.handleLogin()} />;
  }

  handleLogin() {
    // This returns an AuthorizeRequest object. After the request is complete, the object will
    // contain properties detailing the response
    window.amazon.Login.authorize(options, response =>
      this.handleResponse(response)
    );
  }

  handleResponse = authorizationResponse => {
    this.props.updateAuthenticationInfo(authorizationResponse);
  };
}
