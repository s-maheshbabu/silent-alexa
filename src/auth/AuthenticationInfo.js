import IllegalArgumentError from "errors/IllegalArgumentError";
import Cookies from "js-cookie";
import util from "util";

/**
 * Wrapper class to hold access_token and expires_in returned in LoginWithAmazon
 * authorization response.
 * Instances are created only if access_token and expires_in are defined. The token
 * is stored in a cookie called 'amazon_Login_accessToken' because that is the
 * cookie that LoginWithAmazon will look for.
 * https://developer.amazon.com/docs/login-with-amazon/javascript-sdk-reference.html#setUseCookie
 */
// TODO: Rename this to AuthenticationManager now that the class has been refactored to not
// actually hold the authentication information.
class AuthenticationInfo {
  persist(lwaResponse) {
    const numberOfSecondsInADay = 86400;

    if (lwaResponse && lwaResponse.access_token && lwaResponse.expires_in) {
      Cookies.set(AMAZON_LOGIN_COOKIE, lwaResponse.access_token, {
        expires: lwaResponse.expires_in / numberOfSecondsInADay,
        secure: false // TODO: Change localhost to also use https and then change this to true.
      });
    } else {
      const serializedLWAResponse = util.inspect(lwaResponse, {
        showHidden: true,
        depth: null
      });
      throw new IllegalArgumentError(
        `LoginWithAmazon Authorization response is undefined or 
          doesnt have access_token/expires_in.
          lwaResponse: ${serializedLWAResponse}`
      );
    }
  }

  /**
   * @returns true is authentication info is present and false otherwise.
   * This does not guarantee that the tokens haven't already expired.
   * Callers should be able to handle the case of tokens being
   * invalid when they try to use them.
   */
  isPresent = () => {
    return this.getAccessToken() !== undefined;
  };

  /**
   * @returns access_token to access Alexa API
   */
  getAccessToken = () => {
    return undefined;
  };

  /**
   * @returns deleted the cookie that stores the authentication info
   */
  clear = () => {
    Cookies.remove(AMAZON_LOGIN_COOKIE);
  };
}

// Declare a singleton
const instance = new AuthenticationInfo();
Object.freeze(instance);

export default instance;
export const AMAZON_LOGIN_COOKIE = "amazon_Login_accessToken";
