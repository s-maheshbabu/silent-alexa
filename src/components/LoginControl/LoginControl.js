import React, { useContext } from "react";
import HeaderFlatButton from "HeaderFlatButton/HeaderFlatButton";
import { AuthContext } from "auth/AuthContext";

// Options variable to request for implicit grant.
// TODO: Logic for assigning 'deviceSerialNumber' needs to be revisited.
export const options = Object.freeze({
  scope: ["alexa:all", "profile"],
  scope_data: {
    "alexa:all": {
      productID: "Silent_Alexa",
      productInstanceAttributes: { deviceSerialNumber: "12345" }
    }
  },
  popup: false
});

// Redirect path to handle the response from LoginWithAmazon
export const REDIRECT_PATH = "/authresponse";

export default function LoginControl() {
  const { clear, isAuthenticated } = useContext(AuthContext);

  if (isAuthenticated) {
    return <HeaderFlatButton label="Logout" onClick={clear} />;
  } else {
    return <HeaderFlatButton label="Login" onClick={handleLogin} />;
  }
}

function handleLogin() {
  // The authorization service will redirect the user-agent to the redirect path
  // which will contain an authorization response as a URI fragment
  window.amazon.Login.authorize(
    options,
    window.location.href.replace(/\/+$/, "") + REDIRECT_PATH
  );
}
