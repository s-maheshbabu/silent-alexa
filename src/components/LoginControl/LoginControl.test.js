import React from "react";
import { shallow, mount } from "enzyme";
import { options, redirectUri, default as LoginControl } from "./LoginControl";
import AuthenticationInfo from "AuthenticationInfo";
import util from "util";

let loginControl;
let loginControlInstance;
const mockClearAuthenticationInfo = jest.fn();
const mockLWAModule = jest.fn();
const mockIsAuthenticationInfoValid = jest.fn();

beforeEach(() => {
  jest.resetAllMocks();
  Object.defineProperty(window, "amazon", {
    value: { Login: { authorize: mockLWAModule } },
    writable: true
  });
  loginControl = shallow(
    <LoginControl
      isAuthenticationInfoValid={mockIsAuthenticationInfoValid}
      clearAuthenticationInfo={mockClearAuthenticationInfo}
    />
  );
  loginControlInstance = loginControl.instance();
});

it("renders LoginControl with LoginButton component when isAuthenticationInfoValid returns falsy", () => {
  mockIsAuthenticationInfoValid.mockReturnValueOnce(false);
  const wrapper = shallow(
    <LoginControl isAuthenticationInfoValid={mockIsAuthenticationInfoValid} />
  );
  expect(wrapper).toMatchSnapshot();

  wrapper.unmount();
});

it("renders LoginControl with LogoutButton component when isAuthenticationInfoValid returns truthy", () => {
  mockIsAuthenticationInfoValid.mockReturnValueOnce(true);
  const wrapper = shallow(
    <LoginControl isAuthenticationInfoValid={mockIsAuthenticationInfoValid} />
  );

  expect(wrapper).toMatchSnapshot();

  wrapper.unmount();
});

it("verifies that props are passed to HeaderFlatButton(Login) Component", () => {
  const button = loginControl.find("HeaderFlatButton");
  expect(Object.keys(button.props()).length).toBe(2);

  const labelProp = button.prop("label");
  expect(labelProp).toBe("Login");

  const onClickProp = button.prop("onClick");
  const handleLoginSpy = jest.spyOn(loginControlInstance, "handleLogin");

  onClickProp();

  expect(handleLoginSpy).toHaveBeenCalledTimes(1);
});

it("verifies that props are passed to HeaderFlatButton(Logout) Component", () => {
  mockIsAuthenticationInfoValid.mockReturnValueOnce(true);
  const loginControl = shallow(
    <LoginControl
      isAuthenticationInfoValid={mockIsAuthenticationInfoValid}
      clearAuthenticationInfo={mockClearAuthenticationInfo}
    />
  );

  const button = loginControl.find("HeaderFlatButton");
  expect(Object.keys(button.props()).length).toBe(2);

  const labelProp = button.prop("label");
  expect(labelProp).toBe("Logout");

  const onClickProp = button.prop("onClick");
  const handleLogoutSpy = jest.spyOn(loginControl.instance(), "handleLogout");

  onClickProp();

  expect(handleLogoutSpy).toHaveBeenCalled();
});

it("verifies that clearAuthenticationInfo is called when handleLogout is called", () => {
  loginControlInstance.handleLogout();

  expect(mockClearAuthenticationInfo).toHaveBeenCalledTimes(1);
});

it("verifies that lwa authorization is called when handleLogin is called", () => {
  loginControlInstance.handleLogin();

  expect(mockLWAModule).toHaveBeenCalledTimes(1);
  expect(mockLWAModule.mock.calls[0][0]).toBe(options);
  expect(mockLWAModule.mock.calls[0][1]).toBe(redirectUri);
});
