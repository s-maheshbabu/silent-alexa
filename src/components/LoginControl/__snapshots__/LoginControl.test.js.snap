// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders LoginControl with LoginButton component when the user is not authenticated 1`] = `
<DocumentFragment>
  <button
    class="header-flat-button"
    style="border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: 0px; padding: 0px; outline: none; font-size: inherit; font-weight: inherit; position: relative; height: 36px; line-height: 36px; min-width: 88px; color: rgba(0, 0, 0, 0.87); transition: all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms; border-radius: 2px; user-select: none; overflow: hidden; background-color: rgba(0, 0, 0, 0); text-align: center;"
    tabindex="0"
    type="button"
  >
    <div>
      <span
        style="position: relative; padding-left: 16px; padding-right: 16px; vertical-align: middle; letter-spacing: 0; text-transform: uppercase; font-weight: 500; font-size: 14px;"
      >
        Login
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders LoginControl with LogoutButton component when the user is authenticated 1`] = `
<DocumentFragment>
  <button
    class="header-flat-button"
    style="border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: 0px; padding: 0px; outline: none; font-size: inherit; font-weight: inherit; position: relative; height: 36px; line-height: 36px; min-width: 88px; color: rgba(0, 0, 0, 0.87); transition: all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms; border-radius: 2px; user-select: none; overflow: hidden; background-color: rgba(0, 0, 0, 0); text-align: center;"
    tabindex="0"
    type="button"
  >
    <div>
      <span
        style="position: relative; padding-left: 16px; padding-right: 16px; vertical-align: middle; letter-spacing: 0; text-transform: uppercase; font-weight: 500; font-size: 14px;"
      >
        Logout
      </span>
    </div>
  </button>
</DocumentFragment>
`;
